<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Raw-data-overview" data-toc-modified-id="Raw-data-overview-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Raw data overview</a></span></li><li><span><a href="#Data-cleaned-from-special-characters-on-NaN" data-toc-modified-id="Data-cleaned-from-special-characters-on-NaN-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data cleaned from special characters on NaN</a></span><ul class="toc-item"><li><span><a href="#Import-data-and-check-for-misssing-values-in-one-value-(NaN)" data-toc-modified-id="Import-data-and-check-for-misssing-values-in-one-value-(NaN)-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Import data and check for misssing values in one value (NaN)</a></span></li><li><span><a href="#Changing-name-of-columns" data-toc-modified-id="Changing-name-of-columns-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Changing name of columns</a></span></li><li><span><a href="#Creating-new-variable-Time-and-Date-and-formating-it" data-toc-modified-id="Creating-new-variable-Time-and-Date-and-formating-it-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Creating new variable Time and Date and formating it</a></span><ul class="toc-item"><li><span><a href="#Reformat-time,-date-variables-and-create-a-Week-Day-column" data-toc-modified-id="Reformat-time,-date-variables-and-create-a-Week-Day-column-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Reformat time, date variables and create a Week Day column</a></span></li></ul></li><li><span><a href="#Cleaning-colum-Station-to-remove-Barcelona" data-toc-modified-id="Cleaning-colum-Station-to-remove-Barcelona-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Cleaning colum Station to remove Barcelona</a></span></li><li><span><a href="#Droping-column-Latitude,-Longitude-and-Generated" data-toc-modified-id="Droping-column-Latitude,-Longitude-and-Generated-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Droping column Latitude, Longitude and Generated</a></span><ul class="toc-item"><li><span><a href="#Create-a-subset-to-keep-information-about-location-of-the-stations-and-export-it" data-toc-modified-id="Create-a-subset-to-keep-information-about-location-of-the-stations-and-export-it-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Create a subset to keep information about location of the stations and export it</a></span></li><li><span><a href="#Drop-colums-Latitude,-Longitude-and-Generated" data-toc-modified-id="Drop-colums-Latitude,-Longitude-and-Generated-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Drop colums Latitude, Longitude and Generated</a></span></li></ul></li><li><span><a href="#Drop-duplicates" data-toc-modified-id="Drop-duplicates-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Drop duplicates</a></span></li><li><span><a href="#What-we-do-with-the-missing-values?" data-toc-modified-id="What-we-do-with-the-missing-values?-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>What we do with the missing values?</a></span><ul class="toc-item"><li><span><a href="#NaN-in-variables-related-with-time" data-toc-modified-id="NaN-in-variables-related-with-time-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>NaN in variables related with time</a></span><ul class="toc-item"><li><span><a href="#Check-all-variables-related-to-time-together" data-toc-modified-id="Check-all-variables-related-to-time-together-2.7.1.1"><span class="toc-item-num">2.7.1.1&nbsp;&nbsp;</span>Check all variables related to time together</a></span></li><li><span><a href="#Delete-columns-of-the-hour-taken-the-measure-and-keep-the-generated-time" data-toc-modified-id="Delete-columns-of-the-hour-taken-the-measure-and-keep-the-generated-time-2.7.1.2"><span class="toc-item-num">2.7.1.2&nbsp;&nbsp;</span>Delete columns of the hour taken the measure and keep the generated time</a></span></li></ul></li><li><span><a href="#Checking-again-our-NaN-table" data-toc-modified-id="Checking-again-our-NaN-table-2.7.2"><span class="toc-item-num">2.7.2&nbsp;&nbsp;</span>Checking again our NaN table</a></span></li></ul></li></ul></li><li><span><a href="#Export-data" data-toc-modified-id="Export-data-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Export data</a></span></li></ul></div>

# Project Week 2: The quality of air in Barcelona (November 2018)- CELANING THE DATA

## Raw data overview

In [50]:
import pandas as pd
import numpy as np
import os

In [51]:
air_quality = pd.read_csv("data/air-quality-nov-2018.csv")

In [52]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5744 entries, 0 to 5743
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Station       5744 non-null   object 
 1   Air Quality   5744 non-null   object 
 2   Longitude     5744 non-null   float64
 3   Latitude      5744 non-null   float64
 4   O3 Hour       4268 non-null   object 
 5   O3 Quality    4268 non-null   object 
 6   O3 Value      4101 non-null   float64
 7   NO2 Hour      5689 non-null   object 
 8   NO2 Quality   5689 non-null   object 
 9   NO2 Value     5460 non-null   float64
 10  PM10 Hour     3722 non-null   object 
 11  PM10 Quality  3722 non-null   object 
 12  PM10 Value    3647 non-null   float64
 13  Generated     5744 non-null   object 
 14  Date Time     5744 non-null   int64  
dtypes: float64(5), int64(1), object(9)
memory usage: 673.2+ KB


This data set has 5744 entries in total. As a descriptive variable we will consider Station to be used to display the missing values per column per station. We have seen already looking a bit trought de data that might a pattern of missing values for certain stations and certain pollutants.

We are gonna display a table with missing values x column and x station.

In [53]:
#missing values in station x variable
nan_sants=air_quality[air_quality["Station"]=='Barcelona - Sants'].isnull().sum()
nan_eixample=air_quality[air_quality["Station"]=='Barcelona - Eixample'].isnull().sum()
nan_gracia=air_quality[air_quality["Station"]=='Barcelona - Gràcia'].isnull().sum()
nan_ciutadella=air_quality[air_quality["Station"]=='Barcelona - Ciutadella'].isnull().sum()
nan_valldhebron=air_quality[air_quality["Station"]=='Barcelona - Vall Hebron'].isnull().sum()
nan_palaureial=air_quality[air_quality["Station"]=='Barcelona - Palau Reial'].isnull().sum()
nan_poblenou=air_quality[air_quality["Station"]=='Barcelona - Poblenou'].isnull().sum()
nan_fabra=air_quality[air_quality["Station"]=='Barcelona - Observ Fabra'].isnull().sum()

In [54]:
nan_x_station= [nan_sants,nan_eixample,nan_gracia,nan_ciutadella,nan_valldhebron,nan_palaureial,nan_poblenou,nan_fabra]

stations = list(air_quality["Station"].unique())

nan_matrix_x_column_x_station= pd.concat(nan_x_station, axis=1)

nan_matrix_x_column_x_station.columns=stations
nan_matrix_x_column_x_station = nan_matrix_x_column_x_station.drop("Station")

nan_matrix_x_column_x_station =nan_matrix_x_column_x_station.transpose()
nan_matrix_x_column_x_station

Air Quality  Longitude  Latitude  O3 Hour  \
Barcelona - Sants                   0          0         0      718   
Barcelona - Eixample                0          0         0        7   
Barcelona - Gràcia                  0          0         0       11   
Barcelona - Ciutadella              0          0         0        4   
Barcelona - Vall Hebron             0          0         0        6   
Barcelona - Palau Reial             0          0         0        7   
Barcelona - Poblenou                0          0         0      718   
Barcelona - Observ Fabra            0          0         0        5   

                          O3 Quality  O3 Value  NO2 Hour  NO2 Quality  \
Barcelona - Sants                718       718         7            7   
Barcelona - Eixample               7        73         7            7   
Barcelona - Gràcia                11        74        11           11   
Barcelona - Ciutadella             4        17         4            4   
Barcelona - Vall Hebron            6         7         6            6   
Barcelona - Palau Reial            7        17         7            7   
Barcelona - Poblenou             718       718         8            8   
Barcelona - Observ Fabra           5        19         5            5   

                          NO2 Value  PM10 Hour  PM10 Quality  PM10 Value  \
Barcelona - Sants                47        718           718         718   
Barcelona - Eixample             74          0             0           0   
Barcelona - Gràcia               74        507           507         507   
Barcelona - Ciutadella           17        718           718         718   
Barcelona - Vall Hebron           7         73            73          73   
Barcelona - Palau Reial          17          0             0          22   
Barcelona - Poblenou             29          2             2          34   
Barcelona - Observ Fabra         19          4             4          25   

                          Generated  Date Time  
Barcelona - Sants                 0          0  
Barcelona - Eixample              0          0  
Barcelona - Gràcia                0          0  
Barcelona - Ciutadella            0          0  
Barcelona - Vall Hebron           0          0  
Barcelona - Palau Reial           0          0  
Barcelona - Poblenou              0          0  
Barcelona - Observ Fabra          0          0

If we look, for instance, to Gràcia in the O3 columns, it shows 11 missing values per hour and quality and 74 for value. That makes no sense as if there is a measure of one pollutant we should be able to see the same amount of missing value for the three variables.

Looking at this on the data set, we realized that certain entries are consider as values and they should be considered as mising values.

Those are the characters considered as values and not missing values:<b>
  
**"n/a","na","--"**

## Data cleaned from special characters on NaN

### Import data and check for misssing values in one value (NaN)

This time we will import the initial dataset considering those characters as missing values.

In [55]:
#Some of the missing values where looking like this characters
missing_values= ["n/a","na","--"]

In [56]:
air_quality = pd.read_csv("data/air-quality-nov-2018.csv",na_values=missing_values)

In [57]:
air_quality.head(15)

Station Air Quality  Longitude  Latitude O3 Hour  \
0          Barcelona - Sants        Good     2.1331   41.3788     NaN   
1       Barcelona - Eixample    Moderate     2.1538   41.3853      0h   
2         Barcelona - Gràcia        Good     2.1534   41.3987      0h   
3     Barcelona - Ciutadella        Good     2.1874   41.3864      0h   
4    Barcelona - Vall Hebron        Good     2.1480   41.4261      0h   
5    Barcelona - Palau Reial        Good     2.1151   41.3875     23h   
6       Barcelona - Poblenou        Good     2.2045   41.4039     NaN   
7   Barcelona - Observ Fabra        Good     2.1239   41.4183     23h   
8          Barcelona - Sants        Good     2.1331   41.3788     NaN   
9       Barcelona - Eixample        Good     2.1538   41.3853      0h   
10        Barcelona - Gràcia        Good     2.1534   41.3987      0h   
11    Barcelona - Ciutadella        Good     2.1874   41.3864      0h   
12   Barcelona - Vall Hebron        Good     2.1480   41.4261      0h   
13   Barcelona - Palau Reial        Good     2.1151   41.3875      0h   
14      Barcelona - Poblenou        Good     2.2045   41.4039     NaN   

   O3 Quality  O3 Value NO2 Hour NO2 Quality  NO2 Value PM10 Hour  \
0         NaN       NaN       0h        Good       84.0       NaN   
1        Good       1.0       0h    Moderate      113.0        0h   
2        Good      10.0       0h        Good       73.0       NaN   
3        Good       2.0       0h        Good       86.0       NaN   
4        Good       7.0       0h        Good       69.0       NaN   
5        Good      11.0      23h        Good       57.0       23h   
6         NaN       NaN      23h        Good       86.0       23h   
7        Good      58.0      23h        Good        3.0       23h   
8         NaN       NaN       0h        Good       62.0       NaN   
9        Good       6.0       0h        Good       80.0        1h   
10       Good      21.0       0h        Good       59.0       NaN   
11       Good       4.0       0h        Good       78.0       NaN   
12       Good      29.0       0h        Good       45.0       NaN   
13       Good      27.0       0h        Good       38.0        1h   
14        NaN       NaN       0h        Good       75.0        1h   

   PM10 Quality  PM10 Value        Generated   Date Time  
0           NaN         NaN  01/11/2018 0:00  1541027104  
1          Good        36.0  01/11/2018 0:00  1541027104  
2           NaN         NaN  01/11/2018 0:00  1541027104  
3           NaN         NaN  01/11/2018 0:00  1541027104  
4           NaN         NaN  01/11/2018 0:00  1541027104  
5          Good        23.0  01/11/2018 0:00  1541027104  
6          Good        32.0  01/11/2018 0:00  1541027104  
7          Good        25.0  01/11/2018 0:00  1541027104  
8           NaN         NaN  01/11/2018 1:00  1541030725  
9          Good        35.0  01/11/2018 1:00  1541030725  
10          NaN         NaN  01/11/2018 1:00  1541030725  
11          NaN         NaN  01/11/2018 1:00  1541030725  
12          NaN         NaN  01/11/2018 1:00  1541030725  
13         Good        24.0  01/11/2018 1:00  1541030725  
14         Good        32.0  01/11/2018 1:00  1541030725

In [9]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5744 entries, 0 to 5743
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Station       5744 non-null   object 
 1   Air Quality   5583 non-null   object 
 2   Longitude     5744 non-null   float64
 3   Latitude      5744 non-null   float64
 4   O3 Hour       4268 non-null   object 
 5   O3 Quality    4101 non-null   object 
 6   O3 Value      4101 non-null   float64
 7   NO2 Hour      5689 non-null   object 
 8   NO2 Quality   5460 non-null   object 
 9   NO2 Value     5460 non-null   float64
 10  PM10 Hour     3722 non-null   object 
 11  PM10 Quality  3647 non-null   object 
 12  PM10 Value    3647 non-null   float64
 13  Generated     5744 non-null   object 
 14  Date Time     5744 non-null   int64  
dtypes: float64(5), int64(1), object(9)
memory usage: 673.2+ KB


In [58]:
air_quality["Station"].unique()

array(['Barcelona - Sants', 'Barcelona - Eixample', 'Barcelona - Gràcia',
       'Barcelona - Ciutadella', 'Barcelona - Vall Hebron',
       'Barcelona - Palau Reial', 'Barcelona - Poblenou',
       'Barcelona - Observ Fabra'], dtype=object)

In [59]:
#We still have a lot of null values but we will go more in details later
air_quality.isnull().sum()

Station            0
Air Quality      161
Longitude          0
Latitude           0
O3 Hour         1476
O3 Quality      1643
O3 Value        1643
NO2 Hour          55
NO2 Quality      284
NO2 Value        284
PM10 Hour       2022
PM10 Quality    2097
PM10 Value      2097
Generated          0
Date Time          0
dtype: int64

### Changing name of columns

In [60]:
air_quality = air_quality.rename(columns={"O3 Value":"O3 µg/m3","NO2 Value":"NO2 µg/m3","PM10 Value":"PM10 µg/m3"})

In [61]:
air_quality.head()

Station Air Quality  Longitude  Latitude O3 Hour  \
0        Barcelona - Sants        Good     2.1331   41.3788     NaN   
1     Barcelona - Eixample    Moderate     2.1538   41.3853      0h   
2       Barcelona - Gràcia        Good     2.1534   41.3987      0h   
3   Barcelona - Ciutadella        Good     2.1874   41.3864      0h   
4  Barcelona - Vall Hebron        Good     2.1480   41.4261      0h   

  O3 Quality  O3 µg/m3 NO2 Hour NO2 Quality  NO2 µg/m3 PM10 Hour PM10 Quality  \
0        NaN       NaN       0h        Good       84.0       NaN          NaN   
1       Good       1.0       0h    Moderate      113.0        0h         Good   
2       Good      10.0       0h        Good       73.0       NaN          NaN   
3       Good       2.0       0h        Good       86.0       NaN          NaN   
4       Good       7.0       0h        Good       69.0       NaN          NaN   

   PM10 µg/m3        Generated   Date Time  
0         NaN  01/11/2018 0:00  1541027104  
1        36.0  01/11/2018 0:00  1541027104  
2         NaN  01/11/2018 0:00  1541027104  
3         NaN  01/11/2018 0:00  1541027104  
4         NaN  01/11/2018 0:00  1541027104

### Creating new variable Time and Date and formating it

In [62]:
air_quality[['Generated Date','Generated Time']] = air_quality.Generated.str.split(expand=True)
air_quality

Station Air Quality  Longitude  Latitude O3 Hour  \
0            Barcelona - Sants        Good     2.1331   41.3788     NaN   
1         Barcelona - Eixample    Moderate     2.1538   41.3853      0h   
2           Barcelona - Gràcia        Good     2.1534   41.3987      0h   
3       Barcelona - Ciutadella        Good     2.1874   41.3864      0h   
4      Barcelona - Vall Hebron        Good     2.1480   41.4261      0h   
...                        ...         ...        ...       ...     ...   
5739    Barcelona - Ciutadella        Good     2.1874   41.3864     22h   
5740   Barcelona - Vall Hebron        Good     2.1480   41.4261     22h   
5741   Barcelona - Palau Reial        Good     2.1151   41.3875     22h   
5742      Barcelona - Poblenou        Good     2.2045   41.4039     NaN   
5743  Barcelona - Observ Fabra        Good     2.1239   41.4183     22h   

     O3 Quality  O3 µg/m3 NO2 Hour NO2 Quality  NO2 µg/m3 PM10 Hour  \
0           NaN       NaN       0h        Good       84.0       NaN   
1          Good       1.0       0h    Moderate      113.0        0h   
2          Good      10.0       0h        Good       73.0       NaN   
3          Good       2.0       0h        Good       86.0       NaN   
4          Good       7.0       0h        Good       69.0       NaN   
...         ...       ...      ...         ...        ...       ...   
5739       Good      10.0      22h        Good       57.0       NaN   
5740       Good      32.0      22h        Good       31.0       22h   
5741       Good      40.0      22h        Good       20.0       22h   
5742        NaN       NaN      22h        Good       70.0       22h   
5743       Good      64.0      22h        Good       21.0       22h   

     PM10 Quality  PM10 µg/m3         Generated   Date Time Generated Date  \
0             NaN         NaN   01/11/2018 0:00  1541027104     01/11/2018   
1            Good        36.0   01/11/2018 0:00  1541027104     01/11/2018   
2             NaN         NaN   01/11/2018 0:00  1541027104     01/11/2018   
3             NaN         NaN   01/11/2018 0:00  1541027104     01/11/2018   
4             NaN         NaN   01/11/2018 0:00  1541027104     01/11/2018   
...           ...         ...               ...         ...            ...   
5739          NaN         NaN  30/11/2018 23:00  1543615502     30/11/2018   
5740         Good        21.0  30/11/2018 23:00  1543615502     30/11/2018   
5741         Good        15.0  30/11/2018 23:00  1543615502     30/11/2018   
5742         Good        25.0  30/11/2018 23:00  1543615502     30/11/2018   
5743         Good        12.0  30/11/2018 23:00  1543615502     30/11/2018   

     Generated Time  
0              0:00  
1              0:00  
2              0:00  
3              0:00  
4              0:00  
...             ...  
5739          23:00  
5740          23:00  
5741          23:00  
5742          23:00  
5743          23:00  

[5744 rows x 17 columns]

#### Reformat time, date variables and create a Week Day column

In [63]:
air_quality['Generated Date'] = pd.to_datetime(air_quality['Generated Date'],format='%d/%m/%Y')

In [64]:
air_quality["Week Day"] = air_quality["Generated Date"].dt.day_name()

In [65]:
air_quality['Generated Date']= air_quality['Generated Date'].dt.strftime("%d/%m/%y")

In [66]:
air_quality['Generated Date'] = pd.to_datetime(air_quality['Generated Date'],format='%d/%m/%y')

In [67]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5744 entries, 0 to 5743
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Station         5744 non-null   object        
 1   Air Quality     5583 non-null   object        
 2   Longitude       5744 non-null   float64       
 3   Latitude        5744 non-null   float64       
 4   O3 Hour         4268 non-null   object        
 5   O3 Quality      4101 non-null   object        
 6   O3 µg/m3        4101 non-null   float64       
 7   NO2 Hour        5689 non-null   object        
 8   NO2 Quality     5460 non-null   object        
 9   NO2 µg/m3       5460 non-null   float64       
 10  PM10 Hour       3722 non-null   object        
 11  PM10 Quality    3647 non-null   object        
 12  PM10 µg/m3      3647 non-null   float64       
 13  Generated       5744 non-null   object        
 14  Date Time       5744 non-null   int64         
 15  Gene

In [68]:
air_quality['Generated Time'] = pd.to_datetime(air_quality['Generated Time'],format="%H:%M")

In [69]:
air_quality['Generated Time'] =air_quality['Generated Time'].dt.strftime("%H:%M")

In [70]:
air_quality.head()

Station Air Quality  Longitude  Latitude O3 Hour  \
0        Barcelona - Sants        Good     2.1331   41.3788     NaN   
1     Barcelona - Eixample    Moderate     2.1538   41.3853      0h   
2       Barcelona - Gràcia        Good     2.1534   41.3987      0h   
3   Barcelona - Ciutadella        Good     2.1874   41.3864      0h   
4  Barcelona - Vall Hebron        Good     2.1480   41.4261      0h   

  O3 Quality  O3 µg/m3 NO2 Hour NO2 Quality  NO2 µg/m3 PM10 Hour PM10 Quality  \
0        NaN       NaN       0h        Good       84.0       NaN          NaN   
1       Good       1.0       0h    Moderate      113.0        0h         Good   
2       Good      10.0       0h        Good       73.0       NaN          NaN   
3       Good       2.0       0h        Good       86.0       NaN          NaN   
4       Good       7.0       0h        Good       69.0       NaN          NaN   

   PM10 µg/m3        Generated   Date Time Generated Date Generated Time  \
0         NaN  01/11/2018 0:00  1541027104     2018-11-01          00:00   
1        36.0  01/11/2018 0:00  1541027104     2018-11-01          00:00   
2         NaN  01/11/2018 0:00  1541027104     2018-11-01          00:00   
3         NaN  01/11/2018 0:00  1541027104     2018-11-01          00:00   
4         NaN  01/11/2018 0:00  1541027104     2018-11-01          00:00   

   Week Day  
0  Thursday  
1  Thursday  
2  Thursday  
3  Thursday  
4  Thursday

### Cleaning colum Station to remove Barcelona

In [71]:
air_quality["Station"] = air_quality["Station"].str.replace('Barcelona - ', "")

In [72]:
air_quality.head()

Station Air Quality  Longitude  Latitude O3 Hour O3 Quality  O3 µg/m3  \
0        Sants        Good     2.1331   41.3788     NaN        NaN       NaN   
1     Eixample    Moderate     2.1538   41.3853      0h       Good       1.0   
2       Gràcia        Good     2.1534   41.3987      0h       Good      10.0   
3   Ciutadella        Good     2.1874   41.3864      0h       Good       2.0   
4  Vall Hebron        Good     2.1480   41.4261      0h       Good       7.0   

  NO2 Hour NO2 Quality  NO2 µg/m3 PM10 Hour PM10 Quality  PM10 µg/m3  \
0       0h        Good       84.0       NaN          NaN         NaN   
1       0h    Moderate      113.0        0h         Good        36.0   
2       0h        Good       73.0       NaN          NaN         NaN   
3       0h        Good       86.0       NaN          NaN         NaN   
4       0h        Good       69.0       NaN          NaN         NaN   

         Generated   Date Time Generated Date Generated Time  Week Day  
0  01/11/2018 0:00  1541027104     2018-11-01          00:00  Thursday  
1  01/11/2018 0:00  1541027104     2018-11-01          00:00  Thursday  
2  01/11/2018 0:00  1541027104     2018-11-01          00:00  Thursday  
3  01/11/2018 0:00  1541027104     2018-11-01          00:00  Thursday  
4  01/11/2018 0:00  1541027104     2018-11-01          00:00  Thursday

In [73]:
air_quality['Station'].unique()

array(['Sants', 'Eixample', 'Gràcia', 'Ciutadella', 'Vall Hebron',
       'Palau Reial', 'Poblenou', 'Observ Fabra'], dtype=object)

### Droping column Latitude, Longitude and Generated

#### Create a subset to keep information about location of the stations and export it

In [74]:
station_locations= air_quality.loc[:7,["Station","Latitude","Longitude"]]
station_locations

Station  Latitude  Longitude
0         Sants   41.3788     2.1331
1      Eixample   41.3853     2.1538
2        Gràcia   41.3987     2.1534
3    Ciutadella   41.3864     2.1874
4   Vall Hebron   41.4261     2.1480
5   Palau Reial   41.3875     2.1151
6      Poblenou   41.4039     2.2045
7  Observ Fabra   41.4183     2.1239

In [23]:
# path_export= "/Users/lauratll/Documents/Data_Analysis/IronHack/Week_2/Project_week2/Project-Week-2-Barcelona/your-project/data/station_locations.csv"

# station_locations.to_csv(path_or_buf= path_export)

#### Drop colums Latitude, Longitude and Generated

In [75]:
air_quality=air_quality.drop(columns =["Generated","Latitude","Longitude"])
air_quality

Station Air Quality O3 Hour O3 Quality  O3 µg/m3 NO2 Hour  \
0            Sants        Good     NaN        NaN       NaN       0h   
1         Eixample    Moderate      0h       Good       1.0       0h   
2           Gràcia        Good      0h       Good      10.0       0h   
3       Ciutadella        Good      0h       Good       2.0       0h   
4      Vall Hebron        Good      0h       Good       7.0       0h   
...            ...         ...     ...        ...       ...      ...   
5739    Ciutadella        Good     22h       Good      10.0      22h   
5740   Vall Hebron        Good     22h       Good      32.0      22h   
5741   Palau Reial        Good     22h       Good      40.0      22h   
5742      Poblenou        Good     NaN        NaN       NaN      22h   
5743  Observ Fabra        Good     22h       Good      64.0      22h   

     NO2 Quality  NO2 µg/m3 PM10 Hour PM10 Quality  PM10 µg/m3   Date Time  \
0           Good       84.0       NaN          NaN         NaN  1541027104   
1       Moderate      113.0        0h         Good        36.0  1541027104   
2           Good       73.0       NaN          NaN         NaN  1541027104   
3           Good       86.0       NaN          NaN         NaN  1541027104   
4           Good       69.0       NaN          NaN         NaN  1541027104   
...          ...        ...       ...          ...         ...         ...   
5739        Good       57.0       NaN          NaN         NaN  1543615502   
5740        Good       31.0       22h         Good        21.0  1543615502   
5741        Good       20.0       22h         Good        15.0  1543615502   
5742        Good       70.0       22h         Good        25.0  1543615502   
5743        Good       21.0       22h         Good        12.0  1543615502   

     Generated Date Generated Time  Week Day  
0        2018-11-01          00:00  Thursday  
1        2018-11-01          00:00  Thursday  
2        2018-11-01          00:00  Thursday  
3        2018-11-01          00:00  Thursday  
4        2018-11-01          00:00  Thursday  
...             ...            ...       ...  
5739     2018-11-30          23:00    Friday  
5740     2018-11-30          23:00    Friday  
5741     2018-11-30          23:00    Friday  
5742     2018-11-30          23:00    Friday  
5743     2018-11-30          23:00    Friday  

[5744 rows x 15 columns]

### Drop duplicates

In [76]:
air_quality.drop_duplicates(keep=False,inplace=True)

In [77]:
air_quality

Station Air Quality O3 Hour O3 Quality  O3 µg/m3 NO2 Hour  \
0            Sants        Good     NaN        NaN       NaN       0h   
1         Eixample    Moderate      0h       Good       1.0       0h   
2           Gràcia        Good      0h       Good      10.0       0h   
3       Ciutadella        Good      0h       Good       2.0       0h   
4      Vall Hebron        Good      0h       Good       7.0       0h   
...            ...         ...     ...        ...       ...      ...   
5739    Ciutadella        Good     22h       Good      10.0      22h   
5740   Vall Hebron        Good     22h       Good      32.0      22h   
5741   Palau Reial        Good     22h       Good      40.0      22h   
5742      Poblenou        Good     NaN        NaN       NaN      22h   
5743  Observ Fabra        Good     22h       Good      64.0      22h   

     NO2 Quality  NO2 µg/m3 PM10 Hour PM10 Quality  PM10 µg/m3   Date Time  \
0           Good       84.0       NaN          NaN         NaN  1541027104   
1       Moderate      113.0        0h         Good        36.0  1541027104   
2           Good       73.0       NaN          NaN         NaN  1541027104   
3           Good       86.0       NaN          NaN         NaN  1541027104   
4           Good       69.0       NaN          NaN         NaN  1541027104   
...          ...        ...       ...          ...         ...         ...   
5739        Good       57.0       NaN          NaN         NaN  1543615502   
5740        Good       31.0       22h         Good        21.0  1543615502   
5741        Good       20.0       22h         Good        15.0  1543615502   
5742        Good       70.0       22h         Good        25.0  1543615502   
5743        Good       21.0       22h         Good        12.0  1543615502   

     Generated Date Generated Time  Week Day  
0        2018-11-01          00:00  Thursday  
1        2018-11-01          00:00  Thursday  
2        2018-11-01          00:00  Thursday  
3        2018-11-01          00:00  Thursday  
4        2018-11-01          00:00  Thursday  
...             ...            ...       ...  
5739     2018-11-30          23:00    Friday  
5740     2018-11-30          23:00    Friday  
5741     2018-11-30          23:00    Friday  
5742     2018-11-30          23:00    Friday  
5743     2018-11-30          23:00    Friday  

[5744 rows x 15 columns]

### What we do with the missing values?

##### Check all variables related to time together

In [78]:
hour_columns = air_quality[['Station','O3 Hour','NO2 Hour','PM10 Hour','Generated Date','Generated Time']]

In [79]:
hour_columns.head(50)
#we could just pick one column with the generated time

Station O3 Hour NO2 Hour PM10 Hour Generated Date Generated Time
0          Sants     NaN       0h       NaN     2018-11-01          00:00
1       Eixample      0h       0h        0h     2018-11-01          00:00
2         Gràcia      0h       0h       NaN     2018-11-01          00:00
3     Ciutadella      0h       0h       NaN     2018-11-01          00:00
4    Vall Hebron      0h       0h       NaN     2018-11-01          00:00
5    Palau Reial     23h      23h       23h     2018-11-01          00:00
6       Poblenou     NaN      23h       23h     2018-11-01          00:00
7   Observ Fabra     23h      23h       23h     2018-11-01          00:00
8          Sants     NaN       0h       NaN     2018-11-01          01:00
9       Eixample      0h       0h        1h     2018-11-01          01:00
10        Gràcia      0h       0h       NaN     2018-11-01          01:00
11    Ciutadella      0h       0h       NaN     2018-11-01          01:00
12   Vall Hebron      0h       0h       NaN     2018-11-01          01:00
13   Palau Reial      0h       0h        1h     2018-11-01          01:00
14      Poblenou     NaN       0h        1h     2018-11-01          01:00
15  Observ Fabra      0h       0h        1h     2018-11-01          01:00
16         Sants     NaN       1h       NaN     2018-11-01          02:00
17      Eixample      1h       1h        1h     2018-11-01          02:00
18        Gràcia      1h       1h       NaN     2018-11-01          02:00
19    Ciutadella      1h       1h       NaN     2018-11-01          02:00
20   Vall Hebron      1h       1h       NaN     2018-11-01          02:00
21   Palau Reial      1h       1h        2h     2018-11-01          02:00
22      Poblenou     NaN       1h        2h     2018-11-01          02:00
23  Observ Fabra      1h       1h        2h     2018-11-01          02:00
24         Sants     NaN       1h       NaN     2018-11-01          03:00
25      Eixample      2h       2h        3h     2018-11-01          03:00
26        Gràcia      2h       2h       NaN     2018-11-01          03:00
27    Ciutadella      2h       2h       NaN     2018-11-01          03:00
28   Vall Hebron      2h       2h       NaN     2018-11-01          03:00
29   Palau Reial      2h       2h        3h     2018-11-01          03:00
30      Poblenou     NaN       2h        3h     2018-11-01          03:00
31  Observ Fabra      2h       2h        2h     2018-11-01          03:00
32         Sants     NaN       2h       NaN     2018-11-01          04:00
33      Eixample      3h       3h        4h     2018-11-01          04:00
34        Gràcia      3h       3h       NaN     2018-11-01          04:00
35    Ciutadella      3h       3h       NaN     2018-11-01          04:00
36   Vall Hebron      3h       3h       NaN     2018-11-01          04:00
37   Palau Reial      3h       3h        4h     2018-11-01          04:00
38      Poblenou     NaN       3h        4h     2018-11-01          04:00
39  Observ Fabra      3h       3h        4h     2018-11-01          04:00
40         Sants     NaN       2h       NaN     2018-11-01          05:00
41      Eixample      4h       4h        4h     2018-11-01          05:00
42        Gràcia      4h       4h       NaN     2018-11-01          05:00
43    Ciutadella      4h       4h       NaN     2018-11-01          05:00
44   Vall Hebron      4h       4h       NaN     2018-11-01          05:00
45   Palau Reial      4h       4h        5h     2018-11-01          05:00
46      Poblenou     NaN       4h        5h     2018-11-01          05:00
47  Observ Fabra      4h       4h        5h     2018-11-01          05:00
48         Sants     NaN       4h       NaN     2018-11-01          06:00
49      Eixample      5h       5h        6h     2018-11-01          06:00

We decide to keep just one column for the time.

##### Delete columns of the hour taken the measure and keep the generated time

In [80]:
air_quality=air_quality.drop(columns =['O3 Hour','NO2 Hour','PM10 Hour'])
air_quality


Station Air Quality O3 Quality  O3 µg/m3 NO2 Quality  NO2 µg/m3  \
0            Sants        Good        NaN       NaN        Good       84.0   
1         Eixample    Moderate       Good       1.0    Moderate      113.0   
2           Gràcia        Good       Good      10.0        Good       73.0   
3       Ciutadella        Good       Good       2.0        Good       86.0   
4      Vall Hebron        Good       Good       7.0        Good       69.0   
...            ...         ...        ...       ...         ...        ...   
5739    Ciutadella        Good       Good      10.0        Good       57.0   
5740   Vall Hebron        Good       Good      32.0        Good       31.0   
5741   Palau Reial        Good       Good      40.0        Good       20.0   
5742      Poblenou        Good        NaN       NaN        Good       70.0   
5743  Observ Fabra        Good       Good      64.0        Good       21.0   

     PM10 Quality  PM10 µg/m3   Date Time Generated Date Generated Time  \
0             NaN         NaN  1541027104     2018-11-01          00:00   
1            Good        36.0  1541027104     2018-11-01          00:00   
2             NaN         NaN  1541027104     2018-11-01          00:00   
3             NaN         NaN  1541027104     2018-11-01          00:00   
4             NaN         NaN  1541027104     2018-11-01          00:00   
...           ...         ...         ...            ...            ...   
5739          NaN         NaN  1543615502     2018-11-30          23:00   
5740         Good        21.0  1543615502     2018-11-30          23:00   
5741         Good        15.0  1543615502     2018-11-30          23:00   
5742         Good        25.0  1543615502     2018-11-30          23:00   
5743         Good        12.0  1543615502     2018-11-30          23:00   

      Week Day  
0     Thursday  
1     Thursday  
2     Thursday  
3     Thursday  
4     Thursday  
...        ...  
5739    Friday  
5740    Friday  
5741    Friday  
5742    Friday  
5743    Friday  

[5744 rows x 12 columns]

#### Checking again our NaN table

I am gonna check if it is possible to measure the daily mean of each pollutant, per station. If so, we will be able to replace those with this value. But this is not possible as we have no measure for any day in the month for certain pollutants in certain stations.

In [81]:
#missing values in station x variable
nan_sants=air_quality[air_quality["Station"]=="Sants"].isnull().sum()
nan_eixample=air_quality[air_quality["Station"]=='Eixample'].isnull().sum()
nan_gracia=air_quality[air_quality["Station"]=='Gràcia'].isnull().sum()
nan_ciutadella=air_quality[air_quality["Station"]=='Ciutadella'].isnull().sum()
nan_valldhebron=air_quality[air_quality["Station"]=='Vall Hebron'].isnull().sum()
nan_palaureial=air_quality[air_quality["Station"]=='Palau Reial'].isnull().sum()
nan_poblenou=air_quality[air_quality["Station"]=='Poblenou'].isnull().sum()
nan_fabra=air_quality[air_quality["Station"]=='Observ Fabra'].isnull().sum()

In [82]:
nan_x_station= [nan_sants,nan_eixample,nan_gracia,nan_ciutadella,nan_valldhebron,nan_palaureial,nan_poblenou,nan_fabra]

In [83]:
stations = list(air_quality["Station"].unique())

In [84]:
nan_matrix_x_column_x_station= pd.concat(nan_x_station, axis=1)


In [85]:
nan_matrix_x_column_x_station.columns=stations
nan_matrix_x_column_x_station = nan_matrix_x_column_x_station.drop("Station")

In [86]:
nan_matrix_x_column_x_station =nan_matrix_x_column_x_station.transpose()
nan_matrix_x_column_x_station

Air Quality  O3 Quality  O3 µg/m3  NO2 Quality  NO2 µg/m3  \
Sants                  47         718       718           47         47   
Eixample                0          73        73           74         74   
Gràcia                 66          74        74           74         74   
Ciutadella             17          17        17           17         17   
Vall Hebron             2           7         7            7          7   
Palau Reial             5          17        17           17         17   
Poblenou               16         718       718           29         29   
Observ Fabra            8          19        19           19         19   

              PM10 Quality  PM10 µg/m3  Date Time  Generated Date  \
Sants                  718         718          0               0   
Eixample                 0           0          0               0   
Gràcia                 507         507          0               0   
Ciutadella             718         718          0               0   
Vall Hebron             73          73          0               0   
Palau Reial             22          22          0               0   
Poblenou                34          34          0               0   
Observ Fabra            25          25          0               0   

              Generated Time  Week Day  
Sants                      0         0  
Eixample                   0         0  
Gràcia                     0         0  
Ciutadella                 0         0  
Vall Hebron                0         0  
Palau Reial                0         0  
Poblenou                   0         0  
Observ Fabra               0         0

We can leave the data as it is now.

In [87]:
air_quality.head(20)

Station Air Quality O3 Quality  O3 µg/m3 NO2 Quality  NO2 µg/m3  \
0          Sants        Good        NaN       NaN        Good       84.0   
1       Eixample    Moderate       Good       1.0    Moderate      113.0   
2         Gràcia        Good       Good      10.0        Good       73.0   
3     Ciutadella        Good       Good       2.0        Good       86.0   
4    Vall Hebron        Good       Good       7.0        Good       69.0   
5    Palau Reial        Good       Good      11.0        Good       57.0   
6       Poblenou        Good        NaN       NaN        Good       86.0   
7   Observ Fabra        Good       Good      58.0        Good        3.0   
8          Sants        Good        NaN       NaN        Good       62.0   
9       Eixample        Good       Good       6.0        Good       80.0   
10        Gràcia        Good       Good      21.0        Good       59.0   
11    Ciutadella        Good       Good       4.0        Good       78.0   
12   Vall Hebron        Good       Good      29.0        Good       45.0   
13   Palau Reial        Good       Good      27.0        Good       38.0   
14      Poblenou        Good        NaN       NaN        Good       75.0   
15  Observ Fabra        Good       Good      87.0        Good        8.0   
16         Sants        Good        NaN       NaN        Good       46.0   
17      Eixample        Good       Good      14.0        Good       64.0   
18        Gràcia        Good       Good      32.0        Good       41.0   
19    Ciutadella        Good       Good      24.0        Good       56.0   

   PM10 Quality  PM10 µg/m3   Date Time Generated Date Generated Time  \
0           NaN         NaN  1541027104     2018-11-01          00:00   
1          Good        36.0  1541027104     2018-11-01          00:00   
2           NaN         NaN  1541027104     2018-11-01          00:00   
3           NaN         NaN  1541027104     2018-11-01          00:00   
4           NaN         NaN  1541027104     2018-11-01          00:00   
5          Good        23.0  1541027104     2018-11-01          00:00   
6          Good        32.0  1541027104     2018-11-01          00:00   
7          Good        25.0  1541027104     2018-11-01          00:00   
8           NaN         NaN  1541030725     2018-11-01          01:00   
9          Good        35.0  1541030725     2018-11-01          01:00   
10          NaN         NaN  1541030725     2018-11-01          01:00   
11          NaN         NaN  1541030725     2018-11-01          01:00   
12          NaN         NaN  1541030725     2018-11-01          01:00   
13         Good        24.0  1541030725     2018-11-01          01:00   
14         Good        32.0  1541030725     2018-11-01          01:00   
15         Good        25.0  1541030725     2018-11-01          01:00   
16          NaN         NaN  1541034303     2018-11-01          02:00   
17         Good        34.0  1541034303     2018-11-01          02:00   
18          NaN         NaN  1541034303     2018-11-01          02:00   
19          NaN         NaN  1541034303     2018-11-01          02:00   

    Week Day  
0   Thursday  
1   Thursday  
2   Thursday  
3   Thursday  
4   Thursday  
5   Thursday  
6   Thursday  
7   Thursday  
8   Thursday  
9   Thursday  
10  Thursday  
11  Thursday  
12  Thursday  
13  Thursday  
14  Thursday  
15  Thursday  
16  Thursday  
17  Thursday  
18  Thursday  
19  Thursday

## Export data

In [88]:
#path_export= "/Users/lauratll/Documents/Data_Analysis/IronHack/Week_2/Project_week2/Project-Week-2-Barcelona/your-project/data/air_quality_bcn_18_clean.csv"


In [89]:
#air_quality.to_csv(path_or_buf= path_export)